### PR TITLE
Directly write to MutableBuffer in substring

### LINKED ISSUE #1422: Improve substring kernel performance
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently in `substring` kernel, we use a `Vec` to put substrings and then copy strings from it to make a buffer. We can directly write into a buffer using `MutableBuffer` to avoid memory copying.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1422.

# Rationale for this change
 
 

```
substring               time:   [30.604 ms 30.681 ms 30.767 ms]                   
                        change: [-64.758% -64.555% -64.358%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  6 (6.00%) high mild
  7 (7.00%) high severe
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao @alamb

File: arrow/src/util/bench_util.rs [hunk 1]
```diff
@@ -94,6 +94,15 @@ where
 pub fn create_string_array<Offset: StringOffsetSizeTrait>(
     size: usize,
     null_density: f32,
+) -> GenericStringArray<Offset> {
+    create_string_array_with_len(size, null_density, 4)
+}
+
+/// Creates an random (but fixed-seeded) array of a given size, null density and length of string
```
Comment (sunchao) [inline arrow/src/util/bench_util.rs:101]: nit: "a random"

Maybe: "Creates a random (but fixed-seeded) array of strings with given size, null density and length"

**Review by sunchao — APPROVED** [review]
LGTM with one nit

File: arrow/src/util/bench_util.rs (hunk 1 above)
Comment (viirya) [reply arrow/src/util/bench_util.rs]: Fixed.

Comment (viirya) [reply arrow/src/util/bench_util.rs]: Actually I copied from other functions here. 😄

**Comment by sunchao** [comment]
Merged, thanks!

**Comment by viirya** [comment]
Thanks @sunchao !

**Comment by alamb** [comment]
This is a nice change

**Pull request merged**